tests/Common/NoUnusedTest.elm

```diff
@@ -0,0 +1,355 @@
+module Common.NoUnusedTest exposing (..)
+
+import Comment exposing (Comment, CommentType(..))
+import Common.NoUnused as NoUnused
+import Dict
+import Expect
+import Json.Decode as Decode
+import NoUnused.CustomTypeConstructorArgs
+import NoUnused.CustomTypeConstructors
+import NoUnused.Parameters
+import NoUnused.Patterns
+import NoUnused.Variables
+import Review.Rule exposing (Rule)
+import Review.Test
+import Test exposing (Test, describe, test)
+import TestHelper
+
+
+rules : List Rule
+rules =
+    [ NoUnused.CustomTypeConstructors.rule []
+    , NoUnused.CustomTypeConstructorArgs.rule
+    , NoUnused.Parameters.rule
+    , NoUnused.Patterns.rule
+    , NoUnused.Variables.rule
+    ]
+
+
+adequateCode : Test
+adequateCode =
+    test "should not report anything for adequate code" <|
+        \() ->
+            TestHelper.expectNoErrorsForRules rules
+                """
+module TwoFer exposing (twoFer)
+
+twoFer : Maybe String -> String
+twoFer name =
+    "One for "
+        ++ Maybe.withDefault "you" name
+        ++ ", one for me."
+"""
+
+
+customTypeConstructors : Test
+customTypeConstructors =
+    describe "should detect unused type constructors" <|
+        [ test "NoUnused.CustomTypeConstructors base behavior" <|
+            \() ->
+                """
+module TwoFer exposing (..)
+
+type TwoFer = TwoFerUnused
+"""
+                    |> Review.Test.run (NoUnused.CustomTypeConstructors.rule [])
+                    |> Review.Test.expectErrors
+                        [ Review.Test.error
+                            { message = "Type constructor `TwoFerUnused` is not used."
+                            , details = [ "This type constructor is never used. It might be handled everywhere it might appear, but there is no location where this value actually gets created." ]
+                            , under = "TwoFerUnused"
+                            }
+                        ]
+        , test "decoder behavior" <|
```

Comment:
All the decoder tests are very long, and they obscure the other (probably more intent revealing) tests. I think it probably makes sense to keep the decoder tests long, as I don't think there is a sensible way to shorten them.

However, how would you feel about moving them to another file? This should make the primary tests stand out a bit more.

Reply:
In my mind, it's the opposite, the decoder tests are the primary tests, because they are testing our code. 
The rule tests are only here to support the decoder tests (explaining where the json comes from) and as a quick sanity check, and not actually checking that the rules work correctly, because there are so many use cases. For checking those, one should at the hundreds of unit tests on the rules repositories. 

Should I add a comment explaining this?

Reply:
Maybe we can just leave it :) 

For me the decoder tests are hard to read, and they look fragile as they have a lot of very specific hard coded information. As somebody reading the tests, the decoder tests don't give me much confidence (for example, if there was an error in the test and it was passing when it shouldn't, I definitely wouldn't be able to tell). 

However the rule tests are much easier to read, are much more closely linked to what we want the code to do, and they give me a lot of confidence that the code works as it should. the inputs and expected outputs are concise and understandable, and the connection between them is clear. I think the rule tests also test the decoder as well, so are a kind of integration / end to end type test.

I think the two types of tests serve different purposes. The decoder tests are probably useful when debugging / working on the decoders, and the rules tests are good as documentation.

Reply:
> I think the rule tests also test the decoder as well, so are a kind of integration / end to end type test.

I wish that was the case, unfortunately it's not so.
The reason is that the process of getting the final analysis is split in two: 1) get `elm-review` to apply the rules and 2) get the output and transform it into Exercism format. The decoders are only used in step 2) and unfortunately there is no way to do the 1) -> 2) transition in a pure way, the `elm-review` output can only be produced by the node application, so we cannot have a fully integrated test.

I agree that the decoder tests are hard to read, but I couldn't find a better way. They aren't much more than a sanity check, thankfully all in all the decoders are actually pretty simple, they just read a couple of fields and combine them.